pages/docs/couple-your-code/couple-your-code-gradient-data.md

```diff
@@ -0,0 +1,161 @@
+permalink: couple-your-code-gradient-data.html
+keywords: api, adapter, mapping, gradient, nearest-neighbor-gradient
+summary: "So far, our mesh contains only data. This is sufficient for most of the numerical methods that preCICE offers. For nearest-neighbor-gradient mapping, however, preCICE also requires additional gradient data information. In this step, you learn how to add gradient data to the mesh."
+---
+
+{% version 2.4.0 %}
+This feature is available since version 2.4.0.
+{% endversion %}
+
+When using `nearest-neighbor-gradient` mapping, we require coupling data and additional gradient data. We have seen in [Step 3](couple-your-code-mesh-and-data-access.html) how to write data to the mesh.
+Now, we will learn how to write gradient data to the mesh.
+
+For this purpose, we use the following API methods:
+
+```cpp
+bool isDataGradientRequired(int dataID);
+
+void writeScalarGradientData (
+    int             dataID,
+    int             valueIndex,
+    const double*   gradientValues  );
+
+void writeBlockScalarGradientData (
+      int           dataID,
+      int           size,
+      const int*    valueIndices,
+      const double* gradientValues  );
+
+void writeVectorGradientData (
+    int             dataID,
+    int             valueIndex,
+    const double*   gradientValues,
+    bool            rowsFirst = false );
+
+void writeBlockVectorGradientData (
+      int           dataID,
+      int           size,
+      const int*    valueIndices,
+      const double* gradientValues,
+      bool          rowsFirst = false );
+```
+
+* `isDataGradientRequired` returns a boolean, indicates if the data corresponding to the ID `dataID` requires additional gradient data.
+* `writeScalarGradientData` writes gradient data corresponding to scalar data values, i.e., a vector containing the spatial derivative of a scalar quantity in each space dimension.
+* `ẁriteBlockScalarGradintData` writes multiple scalar gradient data at once (analogue to `writeBlockScalarData`).
+* `writeVectorGradientData` writes gradient data corresponding to vector-valued data, i.e, a Jacobian matrix. The matrix is passed as a 1D-array. The parameter `rowsFirst` allows to select whether the 1D-array contains the data column-wise or row-wise.
+* `ẁriteBlockVectorGradintData` writes multiple vector gradient data at once (analogue to `writeBlockVectorData`)
+
+Let's consider an example for writing block vector gradient data corresponding to the vector data `v0 = (v0x, v0y) , v1 = (v1x, v1y), ... , vn = (vnx, vny)` differentiated in spatial directions x and y.
+Per default, the values are passed as following:
+
+```cpp
+( v0x_dx, v0x_dy, v0y_dx, v0y_dy,
+  v1x_dx, v1x_dy, v1y_dx, v1y_dy,
+  ... ,
+  vnx_dx, vnx_dy, vny_dx, vny_dy  )  
+```
+
+When using the `rowsFirst` parameter, the following format is required:
+
+```cpp
+( v0x_dx, v0y_dx, v1x_dx, v1y_dx, ... , vnx_dx, vny_dx,
+  v0x_dy, v0y_dy, v1x_dy, v1y_dy, ... , vnx_dy, vny_dy  )
+```
```

Comment:
This goes beyond the documentation part, but considering that the 'gradient matrix' is a Jacobian, I think the `rowFirst` argument should be a column first, as the Jacobian contains the derivative in a particular direction in a column, right?

Reply:
> Jacobian contains the derivative in a particular direction in a column, right?

![image](https://user-images.githubusercontent.com/2901384/168245122-5e18ba78-d831-40e7-a848-6694594d269d.png)

yes.
But here, we have a 3D tensor, right? So "rowFirst" or "columnFirst" will never tell the full story, right?